### ISSUE TITLE
Spring-HATEOS: ClassCastException with ControllerLinkBuilder.methodOn() and hot restart

### ISSUE THREAD

**Issue by tryptophane** [spring-projects/spring-boot#8335]
Here is my code for adding a HATEOS-Url to my DTO (gopZusammenfassung):

```java
private void enrichWithHATEOS(GopZusammenfassung gopZusammenfassung, long gopId) {
    gopZusammenfassung.add(linkTo(methodOn(AnmerkungenListeRestController.class)
            .readAnmerkungenListeByGopId(gopId)).withRel("anmerkungen"));
}
```

This works well until a hot restart of the backend gets triggered, e.g. by modifying and saving some code. After that, when I refresh my frontend, I get a ClassCastException involving the proxied controller instance:

```
java.lang.ClassCastException: de.kvbawue.gostamm.pflege.go.api.rest.v1.AnmerkungenListeRestController$$EnhancerByCGLIB$$7ee14ccb cannot be cast to de.kvbawue.gostamm.pflege.go.api.rest.v1.AnmerkungenListeRestController
	at de.kvbawue.gostamm.pflege.go.domain.orchestration.GoPflegeOrchestrationService.enrichWithHATEOS(GoPflegeOrchestrationService.java:164)
	at de.kvbawue.gostamm.pflege.go.domain.orchestration.GoPflegeOrchestrationService.requestGopZusammenfassung(GoPflegeOrchestrationService.java:96)
	at de.kvbawue.gostamm.pflege.go.domain.orchestration.GoPflegeOrchestrationService$$FastClassBySpringCGLIB$$f06dc567.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:721)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:139)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:656)
	at de.kvbawue.gostamm.pflege.go.domain.orchestration.GoPflegeOrchestrationService$$EnhancerBySpringCGLIB$$912aeee0.requestGopZusammenfassung(<generated>)
	at de.kvbawue.gostamm.pflege.go.api.rest.v1.GoPflegeRestController.getGopZusammenfassungByGnrAndQuartal(GoPflegeRestController.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at (...)
```

Similar issues have already been reported long ago and marked as closed  (#3784 and #4117). Some kind of regression maybe?

springBootVersion: 1.5.1.RELEASE
spring-hateoas: 0.23.0.RELEASE
STS Version 3.8.3

**Comment by wilkinsona** [comment]
> Some kind of regression maybe?

Unfortunately, it's impossible to say without knowing more about your application and how its modules are structured. You may need to [customize the restart class loader](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-devtools-customizing-classload). If that doesn't help, can you please provide a small sample that reproduces the problem?

**Comment by tryptophane** [comment]
Thank you wilkinsona and sorry for the late response. I was waiting for my company to open the proxy for gist.github.com, but the mills grind very slowly here, so I will paste the code for my minimal demo application directly into this thread. It' a simple Spring Starter project and reproduces the issue. 

To reproduce it, start the application and open http://localhost:8080/rest/v1/demo in the browser. Everything is fine now. Then modify the code (e.g. introduce a system.out.println() somewhere) and save. The spring-boot-devtools will trigger a restart of the application. If you call the same rest url again, you will get an Exception like the one I pasted into the first post.

I would not know how I could get rid of this problem by customizing the restart class loader...

DemoApplication.java
```java
package com.example;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class DemoApplication {

	public static void main(String[] args) {
		SpringApplication.run(DemoApplication.class, args);
	}
}
```

DemoRestController.java
```java
package com.example;

import static org.springframework.hateoas.mvc.ControllerLinkBuilder.*;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(path = "rest/v1/demo")
public class DemoRestController {

	@RequestMapping(method = RequestMethod.GET)
	public ResponseEntity<?> get() {
		MyDco myDco = new MyDco("field1");
		myDco.add(linkTo(methodOn(DemoRestController.class).linked()).withRel("mylink"));
		return new ResponseEntity<>(myDco, HttpStatus.OK);
	}

	@RequestMapping(path = "/linked", method = RequestMethod.GET)
	public @ResponseBody MyLinkedRes linked() {
		return new MyLinkedRes("foo");
	}

}
```

MyDco.java
```java
package com.example;

import org.springframework.hateoas.ResourceSupport;

public class MyDco extends ResourceSupport {

	private String field1;


	public MyDco(String field1) {
		super();
		this.field1 = field1;
	}

	public String getField1() {
		return field1;
	}

	public void setField1(String field1) {
		this.field1 = field1;
	}

	@Override
	public String toString() {
		return "MyDco [field1=" + field1 + "]";
	}
}
```

MyLinkedRes.java
```java
package com.example;

public class MyLinkedRes {

    String aField;

    public MyLinkedRes(String aField) {
        super();
        this.aField = aField;
    }

    public String getaField() {
        return aField;
    }

    public void setaField(String aField) {
        this.aField = aField;
    }

    @Override
    public String toString() {
        return "MyLinkedRel [aField=" + aField + "]";
    }
}
```

build.gradle
```
buildscript {
	ext {
		springBootVersion = '1.5.1.RELEASE'
	}
	repositories {
		mavenCentral()
	}
	dependencies {
		classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
	}
}

apply plugin: 'java'
apply plugin: 'eclipse'
apply plugin: 'org.springframework.boot'

jar {
	baseName = 'demo'
	version = '0.0.1-SNAPSHOT'
}

sourceCompatibility = 1.8

repositories {
	mavenCentral()
}


dependencies {
	compile('org.springframework.boot:spring-boot-starter')
	compile('org.springframework.boot:spring-boot-devtools')
	compile('org.springframework.boot:spring-boot-starter-web')
	
	compile 'org.springframework.hateoas:spring-hateoas:0.23.0.RELEASE'
	
	testCompile('org.springframework.boot:spring-boot-starter-test')
}
```

**Comment by tryptophane** [comment]
Here is the complete stacktrace:

```
java.lang.ClassCastException: com.example.DemoRestController$$EnhancerByCGLIB$$76e87bcb cannot be cast to com.example.DemoRestController
	at com.example.DemoRestController.get(DemoRestController.java:20) ~[bin/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_60]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_60]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_60]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_60]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:116) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:963) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:897) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:622) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846) ~[spring-webmvc-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:230) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) ~[tomcat-embed-websocket-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:105) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:81) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198) ~[tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:474) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:349) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:783) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:798) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1434) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_60]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-8.5.11.jar:8.5.11]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
```

**Comment by snicoll** [comment]
The question @wilkinsona asked you corresponds exactly to the problem you're having. And the link he gave you describes how to fix it.

**Comment by tryptophane** [comment]
> The question @wilkinsona asked you corresponds exactly to the problem you're having

Which question? @wilkinsona asked me to provide a small sample that reproduces the problem, as I did.

> And the link he gave you describes how to fix it.

No, I don't think it does. Every restart triggered by the restart classloader causes this issue, so I would have to exclude all of my sources, which would be the same as not using spring-devtools at all.

This issue is not invalid, and should not have been closed.

**Comment by snicoll** [comment]
Ah, I got confused by the fact that the stacktrace corresponds exactly to the info Andy gave you. Andy said the following, including a link.

> You may need to customize the restart class loader.

Anyway, if I remove Spring-HATEOS the exception goes away so that's really weird indeed. @olivergierke can you please have a look to it?

**Comment by tryptophane** [comment]
thank you @snicoll :-)

**Comment by wilkinsona** [comment]
Something's caching the `RestartClassLoader`. The class of the object returned from `methodOn` is using one `RestartClassLoader` and `DemoRestController.class` is using another.

**Comment by wilkinsona** [comment]
This should have been taken care of by Devtools' `HateoasObjenesisCacheDisabler` but Spring HATEOAS has made the field in `DummyInvocationUtils` final so setting it reflectively no longer works.

**Comment by wilkinsona** [comment]
This was originally fixed by #3784. We need an alternative as we can't see the field any more (well, we could, but I'm not comfortable with changing the value of a final field using reflection).

**Comment by tryptophane** [comment]
Thank you, I can confirm the issue is now fixed :-)
